api/app/Http/Middleware/AuthMiddleware.php

```diff
@@ -0,0 +1,58 @@
+<?php namespace App\Http\Middleware;
+
+use Closure;
+use Tymon\JWTAuth\JWTAuth;
+use Illuminate\Http\Request;
+use App\Exceptions\ForbiddenException;
+
+class AuthMiddleware
+{
+    /**
+     * JWT Auth
+     *
+     * @var JWTAuth
+     */
+    protected $jwtAuth;
+
+    /**
+     * Assign dependencies.
+     *
+     * @param  JWTAuth  $jwtAuth
+     * @return void
+     */
+    public function __construct(JWTAuth $jwtAuth)
+    {
+        $this->jwtAuth = $jwtAuth;
+    }
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param  Request  $request
+     * @param  Closure  $next
+     * @return mixed
+     */
+    public function handle(Request $request, Closure $next)
+    {
+        // Get possible x amount of parameters passed to the middleware
+        $userTypes = array_slice(func_get_args(), 2);
+
+        $token = $this->jwtAuth->getTokenFromRequest($request);
+        $user = $this->jwtAuth->getUser($token);
+
+        // Check the restriction types
+        if (in_array('admin', $userTypes) and $user->user_type == 'admin') {
```

Comment:
@artstorm yea I wasn't sure about entrust, it was what I came across from a quick google. Take a look into lock and see if it works out for us, I'd rather not build a permission system if someone has already done it.

Our needs can be fairly complex as there may be many ways that a user might be able to access a particular resource; it will pretty much be a (implicit or explicit) HABTM relationship between user permissions and entities. A user might have implicit access to an entity via another one, or explicit rights granted. For these more complex permissions I imagine we will have a model method on the user like `if ($this->jwtAuth->getUser($request)->canAccess($article))` where `$article` implements a `UserPermission` interface which defines the method `userCanAccess($userId);`